### DISCUSSION TITLE
Persistent memory across threads

### DISCUSSION THREAD

**Discussion by anuroop18** [JoshuaC215/agent-service-toolkit#111] [Ideas]
Can we implement a persistent memory across chats using some user_id which we can select in streamlit app.
I am trying to implement it, any ideas are welcome.

I use following ideas:
```Python
from langgraph.store.base import BaseStore
from langgraph.store.memory import InMemoryStore

# using a node to write memory.
async def awrite_memory(state: AgentState, config: RunnableConfig, store: BaseStore) -> AgentState:
    """Reflect on the chat history and save a memory to the store."""
    user_id = config["configurable"].get("user_id", "default_user")
    namespace = ("memory", user_id)
    key = "user_memory"
    existing_memory = store.get(namespace, key)
    existing_memory_content = existing_memory.value.get('memory') if existing_memory else "No existing memory found."

    system_msg = CREATE_MEMORY_INSTRUCTION.format(memory=existing_memory_content)
    m = models[config["configurable"].get("model", "gpt-4o-mini")]
    messages = [SystemMessage(content=system_msg)] + state["messages"]
    response = await m.ainvoke(messages, config)

    # Overwrite the existing memory in the store 
    key = "user_memory"
    store.put(namespace, key,{"memory": response.content})
    return {"messages": state["messages"]}

CREATE_MEMORY_INSTRUCTION = """"You are collecting information about the user to personalize your responses.

CURRENT USER INFORMATION:
{memory}

INSTRUCTIONS:
1. Review the chat history below carefully
2. Identify new information about the user, such as:
   - Personal details (name, location)
   - Preferences (likes, dislikes)
   - Interests and hobbies
   - Past experiences
   - Goals or future plans
3. Merge any new information with existing memory
4. Format the memory as a clear, bulleted list
5. If new information conflicts with existing memory, keep the most recent version

Remember: Only include factual information directly stated by the user. Do not make assumptions or inferences.

Based on the chat history below, please update the user information:"""

# And then passing that memory inside prompt to llm node.
```

But this approcah does not persist my memory over docker restarts. And i am unable to integrate this with streamlit ui. 
Also what response to return from awrite_memory node, if i send no response, i get error.

**Comment by JoshuaC215** [comment]
Awesome! Would love to add a version of this in the main repo if you get it working.

> But this approcah does not persist my memory over docker restarts. 

That's expected since docker restart will clear the memory and you're using an InMemoryStore. You'd need to use another store like Postgres or maybe sqlite to get around that limitation. It's still probably fine for a demo. I wouldn't be able to run it other than InMemory in the hosted demo anyway since it costs $$.

> i am unable to integrate this with streamlit ui.

For integrating with the Streamlit UI, you could add a setting / st.dialog for the user to enter their name or use some `st.query_params.user_id` or similar.

**Reply by anuroop18** [reply]
Thanks for your valuable feedback. I will share when i get it working. 😊

**Reply by JoshuaC215** [reply]
Oh one other thought - I don’t know how comfortable you are with Docker, but for local runs we could use a separate Docker container running a plain Postgres image to serve as the database for a PostgresStore with more persistence.

that might be something to add as a second step once the basic in memory version is working.

**Reply by anuroop18** [reply]
Nice idea. I will apply it. This way the memory will persist across all changes on agent server.
